File: frontend/src/api/family.ts
import { apiRequest } from "./client";

export type FamilyRelationType =
  | "SELF"
  | "FATHER"
  | "MOTHER"
  | "SPOUSE"
  | "CHILD"
  | "SIBLING"
  | "GRANDPARENT"
  | "OTHER";

export type FamilyMemberRole = "OWNER" | "MEMBER" | "GUARDIAN" | "DEPENDENT";
export type FamilyStatus = "ACTIVE" | "REMOVED";
export type FamilyMemberStatus = "ACTIVE" | "INVITED" | "PENDING_UNREGISTERED" | "REMOVED";
export type FamilyInviteStatus = "PENDING" | "ACCEPTED" | "DECLINED" | "EXPIRED" | "CANCELED";

export type FamilyGroup = {
  id: number;
  name: string;
  owner_user_id: number;
  status: FamilyStatus;
  created_at: string;
  updated_at: string;
};

export type FamilyMember = {
  id: number;
  family_id: number;
  user_id: number | null;
  display_name: string;
  phone_number: string | null;
  email: string | null;
  relation_type: FamilyRelationType;
  member_role: FamilyMemberRole;
  status: FamilyMemberStatus;
  is_registered: boolean;
  created_at: string;
  updated_at: string;
};

export type FamilyGroupDetail = FamilyGroup & {
  members: FamilyMember[];
};

export type FamilyInvite = {
  id: number;
  family_id: number;
  inviter_user_id: number;
  invitee_user_id: number | null;
  invitee_email: string | null;
  invitee_phone: string | null;
  relation_type: FamilyRelationType;
  member_role: FamilyMemberRole;
  status: FamilyInviteStatus;
  expires_at: string;
  used_at: string | null;
  created_at: string;
  invite_code: string | null;
};

export type FamilySentInvite = Omit<FamilyInvite, "invite_code">;

export type FamilyInvitePreview = {
  invite_id: number;
  family_id: number;
  family_name: string;
  inviter_display_name: string;
  invitee_email: string | null;
  status: FamilyInviteStatus;
  expires_at: string;
};

export type FamilyShareSetting = {
  id: number;
  family_id: number;
  owner_user_id: number;
  viewer_user_id: number;
  share_health_records: boolean;
  share_analysis_results: boolean;
  share_diet_records: boolean;
  share_medications: boolean;
  share_challenges: boolean;
  share_exam_reports: boolean;
  receive_analysis_alerts: boolean;
  receive_abnormal_value_alerts: boolean;
  receive_medication_alerts: boolean;
  created_at: string;
  updated_at: string;
};

export type CreateFamilyGroupPayload = {
  name: string;
};

export type UpdateFamilyGroupPayload = {
  name?: string;
};

export type AddUnregisteredFamilyMemberPayload = {
  display_name: string;
  relation_type: FamilyRelationType;
  phone_number?: string | null;
  email?: string | null;
};

export type CreateFamilyInvitePayload = {
  invitee_email?: string | null;
  invitee_phone?: string | null;
  invitee_user_id?: number | null;
  relation_type: FamilyRelationType;
  member_role?: FamilyMemberRole;
};

export type FamilyShareSettingUpdatePayload = Partial<
  Pick<
    FamilyShareSetting,
    | "share_health_records"
    | "share_analysis_results"
    | "share_diet_records"
    | "share_medications"
    | "share_challenges"
    | "share_exam_reports"
    | "receive_analysis_alerts"
    | "receive_abnormal_value_alerts"
    | "receive_medication_alerts"
  >
>;

export function createFamilyGroup(payload: CreateFamilyGroupPayload): Promise<FamilyGroup> {
  return apiRequest<FamilyGroup>("/family/groups", {
    method: "POST",
    body: payload,
  });
}

export function listFamilyGroups(): Promise<FamilyGroup[]> {
  return apiRequest<FamilyGroup[]>("/family/groups");
}

export function getFamilyGroup(familyId: number): Promise<FamilyGroupDetail> {
  return apiRequest<FamilyGroupDetail>(`/family/groups/${familyId}`);
}

export function updateFamilyGroup(familyId: number, payload: UpdateFamilyGroupPayload): Promise<FamilyGroup> {
  return apiRequest<FamilyGroup>(`/family/groups/${familyId}`, {
    method: "PATCH",
    body: payload,
  });
}

export function deleteFamilyGroup(familyId: number): Promise<void> {
  return apiRequest<void>(`/family/groups/${familyId}`, {
    method: "DELETE",
  });
}

export function listFamilyMembers(familyId: number): Promise<FamilyMember[]> {
  return apiRequest<FamilyMember[]>(`/family/groups/${familyId}/members`);
}

export function addUnregisteredFamilyMember(
  familyId: number,
  payload: AddUnregisteredFamilyMemberPayload,
): Promise<FamilyMember> {
  return apiRequest<FamilyMember>(`/family/groups/${familyId}/members/unregistered`, {
    method: "POST",
    body: payload,
  });
}

export function removeFamilyMember(memberId: number): Promise<void> {
  return apiRequest<void>(`/family/members/${memberId}`, {
    method: "DELETE",
  });
}

export function createFamilyInvite(familyId: number, payload: CreateFamilyInvitePayload): Promise<FamilyInvite> {
  return apiRequest<FamilyInvite>(`/family/groups/${familyId}/invites`, {
    method: "POST",
    body: payload,
  });
}

export function listFamilyGroupInvites(familyId: number): Promise<FamilySentInvite[]> {
  return apiRequest<FamilySentInvite[]>(`/family/groups/${familyId}/invites`);
}

export function listMyFamilyInvites(): Promise<FamilyInvite[]> {
  return apiRequest<FamilyInvite[]>("/family/invites/me");
}

export function acceptFamilyInvite(inviteId: number): Promise<FamilyMember> {
  return apiRequest<FamilyMember>(`/family/invites/${inviteId}/accept`, {
    method: "POST",
  });
}

export function declineFamilyInvite(inviteId: number): Promise<FamilyInvite> {
  return apiRequest<FamilyInvite>(`/family/invites/${inviteId}/decline`, {
    method: "POST",
  });
}

export function acceptFamilyInviteByCode(code: string): Promise<FamilyMember> {
  return apiRequest<FamilyMember>("/family/invites/code/accept", {
    method: "POST",
    body: { code },
  });
}

export function previewFamilyInviteByCode(inviteCode: string): Promise<FamilyInvitePreview> {
  return apiRequest<FamilyInvitePreview>("/family/invites/code/preview", {
    method: "POST",
    body: { invite_code: inviteCode },
  });
}

export function listFamilyShareSettings(): Promise<FamilyShareSetting[]> {
  return apiRequest<FamilyShareSetting[]>("/family/share-settings");
}

export function listFamilyGroupShareSettings(familyId: number): Promise<FamilyShareSetting[]> {
  return apiRequest<FamilyShareSetting[]>(`/family/groups/${familyId}/share-settings`);
}

export function updateFamilyShareSetting(
  settingId: number,
  payload: FamilyShareSettingUpdatePayload,
): Promise<FamilyShareSetting> {
  return apiRequest<FamilyShareSetting>(`/family/share-settings/${settingId}`, {
    method: "PATCH",
    body: payload,
  });
}

File: app/models/family.py
from enum import StrEnum

from tortoise import fields, models


class FamilyStatus(StrEnum):
    ACTIVE = "ACTIVE"
    REMOVED = "REMOVED"


class FamilyRelationType(StrEnum):
    SELF = "SELF"
    FATHER = "FATHER"
    MOTHER = "MOTHER"
    SPOUSE = "SPOUSE"
    CHILD = "CHILD"
    SIBLING = "SIBLING"
    GRANDPARENT = "GRANDPARENT"
    OTHER = "OTHER"


class FamilyMemberRole(StrEnum):
    OWNER = "OWNER"
    MEMBER = "MEMBER"
    GUARDIAN = "GUARDIAN"
    DEPENDENT = "DEPENDENT"


class FamilyMemberStatus(StrEnum):
    ACTIVE = "ACTIVE"
    INVITED = "INVITED"
    PENDING_UNREGISTERED = "PENDING_UNREGISTERED"
    REMOVED = "REMOVED"


class FamilyInviteStatus(StrEnum):
    PENDING = "PENDING"
    ACCEPTED = "ACCEPTED"
    DECLINED = "DECLINED"
    EXPIRED = "EXPIRED"
    CANCELED = "CANCELED"


class Family(models.Model):
    id = fields.BigIntField(primary_key=True)
    name = fields.CharField(max_length=100)
    owner_user = fields.ForeignKeyField("models.User", related_name="owned_families", on_delete=fields.CASCADE)
    status = fields.CharEnumField(enum_type=FamilyStatus, default=FamilyStatus.ACTIVE)
    created_at = fields.DatetimeField(auto_now_add=True)
    updated_at = fields.DatetimeField(auto_now=True)

    class Meta:
        table = "families"
        indexes = (("owner_user_id",), ("status",))


class FamilyMember(models.Model):
    id = fields.BigIntField(primary_key=True)
    family = fields.ForeignKeyField("models.Family", related_name="members", on_delete=fields.CASCADE)
    user = fields.ForeignKeyField("models.User", related_name="family_members", null=True, on_delete=fields.SET_NULL)
    display_name = fields.CharField(max_length=100)
    phone_number = fields.CharField(max_length=30, null=True)
    email = fields.CharField(max_length=255, null=True)
    relation_type = fields.CharEnumField(enum_type=FamilyRelationType)
    member_role = fields.CharEnumField(enum_type=FamilyMemberRole, default=FamilyMemberRole.MEMBER)
    status = fields.CharEnumField(enum_type=FamilyMemberStatus, default=FamilyMemberStatus.ACTIVE)
    is_registered = fields.BooleanField(default=True)
    created_at = fields.DatetimeField(auto_now_add=True)
    updated_at = fields.DatetimeField(auto_now=True)

    class Meta:
        table = "family_members"
        indexes = (
            ("family_id",),
            ("user_id",),
            ("family_id", "user_id"),
            ("family_id", "status"),
            ("email",),
            ("phone_number",),
        )


class FamilyInvite(models.Model):
    id = fields.BigIntField(primary_key=True)
    family = fields.ForeignKeyField("models.Family", related_name="invites", on_delete=fields.CASCADE)
    inviter_user = fields.ForeignKeyField("models.User", related_name="sent_family_invites", on_delete=fields.CASCADE)
    invitee_user = fields.ForeignKeyField(
        "models.User",
        related_name="received_family_invites",
        null=True,
        on_delete=fields.SET_NULL,
    )
    invitee_email = fields.CharField(max_length=255, null=True)
    invitee_phone = fields.CharField(max_length=30, null=True)
    code_hash = fields.CharField(max_length=128, unique=True)
    relation_type = fields.CharEnumField(enum_type=FamilyRelationType)
    member_role = fields.CharEnumField(enum_type=FamilyMemberRole, default=FamilyMemberRole.MEMBER)
    expires_at = fields.DatetimeField()
    used_at = fields.DatetimeField(null=True)
    status = fields.CharEnumField(enum_type=FamilyInviteStatus, default=FamilyInviteStatus.PENDING)
    created_at = fields.DatetimeField(auto_now_add=True)

    class Meta:
        table = "family_invites"
        indexes = (
            ("family_id",),
            ("inviter_user_id",),
            ("invitee_user_id",),
            ("invitee_email",),
            ("invitee_phone",),
            ("status",),
            ("expires_at",),
        )


class FamilyShareSetting(models.Model):
    id = fields.BigIntField(primary_key=True)
    family = fields.ForeignKeyField("models.Family", related_name="share_settings", on_delete=fields.CASCADE)
    owner_user = fields.ForeignKeyField("models.User", related_name="family_share_owners", on_delete=fields.CASCADE)
    viewer_user = fields.ForeignKeyField("models.User", related_name="family_share_viewers", on_delete=fields.CASCADE)
    share_health_records = fields.BooleanField(default=False)
    share_analysis_results = fields.BooleanField(default=False)
    share_diet_records = fields.BooleanField(default=False)
    share_medications = fields.BooleanField(default=False)
    share_challenges = fields.BooleanField(default=False)
    share_exam_reports = fields.BooleanField(default=False)
    receive_analysis_alerts = fields.BooleanField(default=False)
    receive_abnormal_value_alerts = fields.BooleanField(default=False)
    receive_medication_alerts = fields.BooleanField(default=False)
    created_at = fields.DatetimeField(auto_now_add=True)
    updated_at = fields.DatetimeField(auto_now=True)

    class Meta:
        table = "family_share_settings"
        indexes = (
            ("family_id",),
            ("owner_user_id",),
            ("viewer_user_id",),
            ("family_id", "owner_user_id", "viewer_user_id"),
        )
